### PR TITLE
Add all workflows

### PR DESCRIPTION
**Pull request by mweidling**
This MR introduces the other workflows proposed on the OCR-D website and adds them to the benchmark creation.

The workflow for slower processors had to be altered, though, because ocrd-skimage-denoise threw some error when using PIL. I substituted it with ocrd-cis-ocropy-denoise.

### This MR depends on https://github.com/OCR-D/quiver-back-end/pull/42.

### CODE REVIEW

**Comment by kba** [comment]
> This MR depends on #42.

Can we merge this first, so the difference from this PR as opposed to #42 becomes clear?

> The workflow for slower processors had to be altered, though, because ocrd-skimage-denoise threw some error when using PIL. I substituted it with ocrd-cis-ocropy-denoise.

This really needs to be investigated. I failed to reproduce the issue with different Pillow/Python versions. Can you open an issue in https://github.com/bertsky/ocrd_wrap so we can find out what's going wrong together with @bertsky?

File: workflows/execute_workflows.sh [hunk 1]
```diff
@@ -92,65 +87,72 @@ do
...
+mkdir -p "$WORKSPACE_DIR"/tmp
+for WORKFLOW in "$OCRD_WORKFLOW_DIR"/*ocr.txt.nf
+do
+    WF_NAME=$(echo "$WORKFLOW" | cut -d'.' -f 1 | rev | cut -d'/' -f 1 | rev)
+    echo "Restore OCR-D workspaces from BagIts …"
+    for BAGIT in "$ROOT"/submodules/quiver-data/*.zip
+    do
+        ocrd zip spill "$BAGIT" -d "$WORKSPACE_DIR"/tmp > "$WORKSPACE_DIR"/log.log
+        # this is a workaround for a werid behaviour where 16_ant_complex.ocrd.zip is sometimes not
+        # correctly extracted (unknown reason)
```
Comment (kba) [inline workflows/execute_workflows.sh:117]: Strange - any error messages?

**Review by kba — APPROVED** [review]

**Comment by mweidling** [comment]
> Can we merge this first, so the difference from this PR as opposed to https://github.com/OCR-D/quiver-back-end/pull/42 becomes clear?

Yes, I thought we tend to #42 first, that's why I made the note.

File: workflows/execute_workflows.sh (hunk 1 above)
Comment (kba) [reply workflows/execute_workflows.sh:119]: Weird indeed. Can you open an issue in core so we can debug this?

File: quiver/benchmark_extraction.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
 import re
 from typing import Any, Dict, List, Union
 
-METS = '{http://www.loc.gov/METS/}'
+from .constants import *
```
Comment (kba) [inline quiver/benchmark_extraction.py]: Always do explicit imports, it's more verbose but makes it clear where all the names come from without an IDE.

File: quiver/constants.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+Constants for QuiVer.
+"""
+
+__all__ = [
+    "METS",
+    "OCRD",
+    "QUIVER_MAIN",
+    "RESULTS"
+]
+
```
Comment (kba) [inline quiver/constants.py:11]:
```suggestion
from ocrd_models import NAMESPACES
```

File: quiver/constants.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+METS = '{http://www.loc.gov/METS/}'
```
Comment (kba) [inline quiver/constants.py:12]:
```suggestion
METS = "{NAMESPACES['mets']}"
```

File: quiver/constants.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+OCRD = '{https://ocr-d.de}'
```
Comment (kba) [inline quiver/constants.py:13]:
```suggestion
OCRD = '{NAMESPACES["ocrd"]}'
```

File: tests/test_benchmark_extraction.py [hunk 6]
```diff
@@ -10,41 +14,41 @@
 from quiver.benchmark_extraction import get_gt_workspace
 from quiver.benchmark_extraction import get_workflow
 from quiver.benchmark_extraction import get_workspace
+from quiver.constants import QUIVER_MAIN
+
+WORKSPACE_DIR = 'tests/assets/benchmarking/16_ant_complex/'
```
Comment (kba) [inline tests/test_benchmark_extraction.py:19]:
```suggestion
WORKSPACE_DIR = Path(__file__).parent / 'assets/benchmarking/16_ant_complex/'
```

File: workflows/execute_workflows.sh [hunk 7]
```diff
@@ -5,6 +5,8 @@ WORKFLOW_DIR="$ROOT"/workflows
 OCRD_WORKFLOW_DIR="$WORKFLOW_DIR"/ocrd_workflows
 WORKSPACE_DIR="$WORKFLOW_DIR"/workspaces
 
+set -eou pipefail
```
Comment (kba) [inline workflows/execute_workflows.sh:8]:
```suggestion
set -euo pipefail
```
(because `pipefail` is an argument to `-o`.)

File: quiver/benchmark_extraction.py [hunk 8]
```diff
@@ -8,7 +8,7 @@
 import re
 from typing import Any, Dict, List, Union
 
-from .constants import *
+from quiver.constants import *
```
Comment (kba) [inline quiver/benchmark_extraction.py:11]: Why?

**Review by kba — CHANGES_REQUESTED** [review]
LGTM, only small stuff remains AFAICT.

File: workflows/execute_workflows.sh (hunk 1 above)
Comment (mweidling) [reply workflows/execute_workflows.sh]: I discovered that this is caused somehow while processing the BagIts. I restructured the bash script (i.e. tried to make it less complex) and that solved the issue.

This is somewhat unsatisfactory, but I will not investigate this further unless the problem re-occurs.

File: tests/test_benchmark_extraction.py (hunk 6 above)
Comment (mweidling) [reply tests/test_benchmark_extraction.py]: Done in 44bae974b4165b602abafc8a80c58809f19098dd.

File: quiver/benchmark_extraction.py (hunk 8 above)
Comment (mweidling) [reply quiver/benchmark_extraction.py]: I thought that because `make benchmarks` is executed from the root of the project (and so is `quiver/benchmark_extraction.py`), this is necessary because `constants.py` wouldn't be found. I just learned that Python still finds it.

Addressed in [ced3251](https://github.com/OCR-D/quiver-back-end/pull/43/commits/ced3251ae1b711a1b197d8fc824536491ddf7cb9).

**Comment by mweidling** [comment]
@kba Are we ready to go? :)

File: quiver/constants.py [hunk 9]
```diff
@@ -8,8 +8,8 @@
     "QUIVER_MAIN",
     "RESULTS"
 ]
-
-METS = '{http://www.loc.gov/METS/}'
-OCRD = '{https://ocr-d.de}'
+from ocrd_models.constants import NAMESPACES
+METS = "{" + f"{NAMESPACES['mets']}" + "}"
```
Comment (kba) [inline quiver/constants.py:12]: If you use `+`, the `f'{...}` in the middle seems unnecessary.

File: requirements.txt [hunk 10]
```diff
@@ -1,9 +1,10 @@
 click >=7
 ocrd >= 2.35.0
-pyyaml
+pyaml
```
Comment (kba) [inline requirements.txt:3]:
```suggestion
```

It's actually `pyyaml` but that is already included via `ocrd`.

File: requirements.txt [hunk 11]
```diff
@@ -1,9 +1,10 @@
...
 requests
```
Comment (kba) [inline requirements.txt:4]:
```suggestion
```

Also included in ocrd

File: requirements.txt [hunk 12]
```diff
@@ -1,9 +1,10 @@
...
 pipgrip
 nextflow
 tomli
 fastapi
 uvicorn[standard]
+pytest
```
Comment (kba) [inline requirements.txt:10]: This should not be in the `requirements.txt` but in a `requirements-test.txt` or similar. You don't need it to run the software, just for testing/development.

File: tests/test_benchmark_extraction.py [hunk 13]
```diff
@@ -2,30 +2,26 @@
...
+                                         get_gt_workspace, get_mean_cer,
+                                         get_metrics_for_page, get_no_of_pages,
+                                         get_page_id, get_workflow,
+                                         get_workflow_model,
+                                         get_workflow_steps, get_workspace)
 from quiver.constants import QUIVER_MAIN
 
-WORKSPACE_DIR = 'tests/assets/benchmarking/16_ant_complex'
-METS_PATH = 'tests/assets/benchmarking/16_ant_complex/mets.xml'
+WORKSPACE_DIR = str(Path(__file__).parent) + '/assets/benchmarking/16_ant_complex'
```
Comment (kba) [inline tests/test_benchmark_extraction.py:17]:
```suggestion
WORKSPACE_DIR = str(Path(__file__).parent / 'assets/benchmarking/16_ant_complex')
```

File: workflows/execute_workflows.sh [hunk 14]
```diff
@@ -53,6 +53,25 @@ git -C submodules/oton submodule update --init
...
+pushd submodules/oton
+pip install .
+popd
+echo "OtoN installation done."
+
+echo "Installing quiver-ocrd"
+pip install .
+echo "quiver-ocrd installation done."
+
+quiver-ocrd || {
```
Comment (kba) [inline workflows/execute_workflows.sh:66]:
```suggestion
which quiver-ocrd || {
```

should be enough. Otherwise, I'd do `x --version` or `x --help` because no-arg calls might result in non-zero exit code.

File: workflows/ocrd_workflows/slower_processors_ocr.txt [hunk 15]
```diff
@@ -3,6 +3,6 @@ ocrd process \
   "anybaseocr-crop -I OCR-D-BIN -O OCR-D-CROP" \
   "cis-ocropy-denoise -I OCR-D-CROP -O OCR-D-BIN-DENOISE -P level-of-operation page" \
   "tesserocr-deskew -I OCR-D-BIN-DENOISE -O OCR-D-BIN-DENOISE-DESKEW -P operation_level page" \
-  "tesserocr-segment -I OCR-D-BIN-DENOISE-DESKEW -O OCR-D-SEG -P shrink_polygons true" \
+  "tesserocr-segment -I OCR-D-BIN-DENOISE-DESKEW -O OCR-D-SEG" \
```
Comment (kba) [inline workflows/ocrd_workflows/slower_processors_ocr.txt:6]: just curious: why this change?

**Review by kba — APPROVED** [review]
Some minor stuff, but LGTM in general.

File: quiver/constants.py (hunk 9 above)
Comment (mweidling) [reply quiver/constants.py]: Addressed in [edd81d6](https://github.com/OCR-D/quiver-back-end/pull/43/commits/edd81d68c561147770c715019b539b9d9aa24ad1)

File: workflows/ocrd_workflows/slower_processors_ocr.txt (hunk 15 above)
Comment (mweidling) [reply workflows/ocrd_workflows/slower_processors_ocr.txt]: The parameter leads to the following error:

```bash
Error executing process > 'ocrd_tesserocr_segment_4'

Caused by:
  Process `ocrd_tesserocr_segment_4` terminated with an error exit status (1)

Command executed:

  docker run --rm -u $(id -u) -v /some/path/quiver-back-end/workflows/workspaces/16_ant_complex_slower_processors_ocr/:/ocrd-workspace -v $HOME/ocrd_models:/usr/local/share/ocrd-resources -w /ocrd-workspace -- ocrd/all:maximum ocrd-tesserocr-segment -I OCR-D-BIN-DENOISE-DESKEW -O OCR-D-SEG -P shrink_polygons true

Command exit status:
  1

Command output:
  (empty)

Command error:
  19:11:43.723 INFO processor.TesserocrSegment - INPUT FILE 0 / phys_0008
  19:11:43.774 INFO processor.TesserocrSegment - Page 'phys_0008' images will use 300 DPI from image meta-data
  19:11:43.774 INFO processor.TesserocrSegment - Processing page 'phys_0008'
  19:11:44.066 INFO ocrd.workspace.save_image_file - created file ID: OCR-D-SEG_0008.IMG-BIN, file_grp: OCR-D-SEG, path: OCR-D-SEG/OCR-D-SEG_0008.IMG-BIN.png
  Traceback (most recent call last):
    File "/usr/local/bin/ocrd-tesserocr-segment", line 8, in <module>
      sys.exit(ocrd_tesserocr_segment())
    File "/usr/local/lib/python3.6/dist-packages/click/core.py", line 1128, in __call__
      return self.main(*args, **kwargs)
    File "/usr/local/lib/python3.6/dist-packages/click/core.py", line 1053, in main
      rv = self.invoke(ctx)
    File "/usr/local/lib/python3.6/dist-packages/click/core.py", line 1395, in invoke
      return ctx.invoke(self.callback, **ctx.params)
    File "/usr/local/lib/python3.6/dist-packages/click/core.py", line 754, in invoke
      return __callback(*args, **kwargs)
    File "/usr/local/lib/python3.6/site-packages/ocrd_tesserocr/cli.py", line 18, in ocrd_tesserocr_segment
      return ocrd_cli_wrap_processor(TesserocrSegment, *args, **kwargs)
    File "/build/core/ocrd/ocrd/decorators/__init__.py", line 117, in ocrd_cli_wrap_processor
      run_processor(processorClass, ocrd_tool, mets, workspace=workspace, **kwargs)
    File "/build/core/ocrd/ocrd/processor/helpers.py", line 107, in run_processor
      processor.process()
    File "/usr/local/lib/python3.6/site-packages/ocrd_tesserocr/segment.py", line 69, in process
      return self.recognizer.process()
    File "/usr/local/lib/python3.6/site-packages/ocrd_tesserocr/recognize.py", line 445, in process
      self._process_regions_in_page(tessapi.GetIterator(), page, page_coords, pcgts_mapping, dpi)
    File "/usr/local/lib/python3.6/site-packages/ocrd_tesserocr/recognize.py", line 525, in _process_regions_in_page
      for symbol in iterate_level(it, RIL.SYMBOL, parent=RIL.BLOCK)])
    File "/usr/local/lib/python3.6/site-packages/ocrd_tesserocr/recognize.py", line 1456, in join_polygons
      for poly in polygons]))
    File "/usr/local/lib/python3.6/site-packages/ocrd_tesserocr/recognize.py", line 1456, in <listcomp>
      for poly in polygons]))
  AttributeError: 'list' object has no attribute 'type'
```

I'm using the maximum image that has been created last week (ID 8784a41dc959) which yields `Version 0.16.0, ocrd/core 2.44.0` for `ocrd-tesserocr-segment --version`.
Smuggling it in here was kind of a quick fix, admittedly.

Comment (mweidling) [reply workflows/ocrd_workflows/slower_processors_ocr.txt]: Referenced in a separate ticket.
